src/dft/backends/cufft/forward.cpp

```diff
@@ -0,0 +1,179 @@
+#include "oneapi/mkl/dft/detail/types_impl.hpp"
+#include "oneapi/mkl/dft/detail/descriptor_impl.hpp"
+
+// cuFFT headers
+#include <cuda_runtime.h>
+#include <cufft.h>
+
+namespace oneapi::mkl::dft::cufft {
+namespace detail {
+
+template <dft::detail::precision prec, dft::detail::domain dom>
+inline dft::detail::commit_impl *get_commit(dft::detail::descriptor<prec, dom> &desc) {
+    auto commit_handle = dft::detail::get_commit(desc);
+    if (commit_handle == nullptr || commit_handle->get_backend() != backend::cufft) {
+        throw mkl::invalid_argument("DFT", "compute_backward",
+                                    "DFT descriptor has not been commited for cuFFT");
+    }
+    return commit_handle;
+}
+
+/// Throw an mkl::invalid_argument if the runtime param in the descriptor does not match
+/// the expected value.
+template <dft::detail::config_param Param, dft::detail::config_value Expected, typename DescT>
+inline auto expect_config(DescT &desc, const char *message) {
+    dft::detail::config_value actual{ 0 };
+    desc.get_value(Param, &actual);
+    if (actual != Expected) {
+        throw mkl::invalid_argument("DFT", "compute_forward", message);
+    }
+}
+} // namespace detail
+
+// BUFFER version
+
+//In-place transform
+template <typename descriptor_type, typename data_type>
+ONEMKL_EXPORT void compute_forward(descriptor_type &desc, sycl::buffer<data_type, 1> &inout) {
+    if constexpr (descriptor_type::precision == dft::detail::precision::SINGLE &&
+                  !std::is_floating_point_v<data_type>) {
+        detail::expect_config<dft::detail::config_param::PLACEMENT,
+                              dft::detail::config_value::INPLACE>(desc,
+                                                                  "Unexpected value for placement");
+        auto commit = get_commit(desc);
+        auto queue = commit->get_queue();
+        auto plan = *static_cast<cufftHandle *>(commit->get_handle());
+
+        cufftResult result = CUFFT_NOT_SUPPORTED;
+        cufftResult *result_ptr = &result;
+
+        queue.submit([&](sycl::handler &cgh) {
+            auto inout_acc = inout.template get_access<sycl::access::mode::read_write>(cgh);
+
+            cgh.host_task([=](sycl::interop_handle ih) {
+                auto inout_native = reinterpret_cast<cufftComplex *>(
+                    ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(inout_acc));
+                auto stream = ih.get_native_queue<sycl::backend::ext_oneapi_cuda>();
+                *result_ptr = cufftSetStream(plan, stream);
+                if (*result_ptr == CUFFT_SUCCESS) {
+                    *result_ptr = cufftExecC2C(plan, inout_native, inout_native, CUFFT_FORWARD);
+                }
+            });
+        }).wait();
+
+        if (result != CUFFT_SUCCESS) {
+            throw oneapi::mkl::exception("DFT", "compute_forward(desc, inout)",
+                                         "cuFFTResult value of " + std::to_string(result));
```

Comment:
What stops us from throwing this in the host task, avoiding the need for the `.wait()`? Same goes for `compute_backward`.

Reply:
Nothing apparently so I've done that in https://github.com/oneapi-src/oneMKL/pull/278/commits/548a3d933e1501cbcac7135fec1beeb26724261f. I must have got myself confused into thinking that was awkward at some point.